addons/mail/models/discuss/discuss_channel_member.py

```diff
@@ -212,16 +212,21 @@ def write(self, vals):
     def unlink(self):
         # sudo: discuss.channel.rtc.session - cascade unlink of sessions for self member
         self.sudo().rtc_session_ids.unlink()  # ensure unlink overrides are applied
-        # Always unlink members of sub-channels as well. Member list should be
-        # kept in sync.
-        sub_members = []
-        for parent in self.channel_id.filtered(lambda c: c.sub_channel_ids):
-            sub_members += parent.sub_channel_ids.channel_member_ids.filtered(
-                lambda m: m.partner_id in set(parent.channel_member_ids.partner_id)
-                if m.partner_id
-                else m.guest_id in set(parent.channel_member_ids.guest_id)
+        # always unlink members of sub-channels as well
+        domain = expression.FALSE_DOMAIN
+        for (p_id, g_id), rec in self.grouped(
+                lambda m: (m.partner_id.id, m.guest_id.id)
+        ).items():
+            domain = expression.OR(
+                [
+                    domain,
+                    [
+                        ("partner_id" if p_id else "guest_id", "=", p_id or g_id),
+                        ("channel_id", "in", rec.channel_id.sub_channel_ids.ids),
+                    ],
+                ]
```

Comment:
Now that I see it, I'm wondering if it's worth it to group them, it's probably easier to make subdomains for each member directly.

```suggestion
        domains = [
            [
                ("partner_id", "=", member.partner_id.id),
                ("guest_id", "=", member.guest_id.id),
                ("channel_id", "in", member.channel_id.sub_channel_ids.ids),
            ]
            for member in self
        ]
```